### PR TITLE
test: Check if email is sent for new mentorship relation requests

### LINKED ISSUE #432: Create tests for checking if email is sent for new mentorship relation requests.
## Description
As a developer for backend api,
I need cover most part of the logic of backend,
so that I can be confident that the app works as it is supposed to.

We send email notifications when new mentorship relationship was created. We need to create tests to check the functionality of this specific function. 

PR #276 can help to understand what you may have to use to complete this task

### UPDATE [Required]
- [ ] Write test for checking if the mail was sent if the mentorship request is created.
- [ ] Write test for checking if the mail wasn't sent if the mentorship request wasn't created (Invalid).

### Definition of Done
- [ ] All the required items are completed.
- [ ] Approval by 1 mentor.

### Estimation
12 hours

### PR DESCRIPTION
**Pull request by gaurivn**
### Description
Testing whether email is sent for valid mentorship relation requests and whether email is not sent for invalid mentorship relation requests with 2 added unit tests. Changed the version of MarkupSafe (to 1.1.1) in requirements.txt.

Fixes #432 

### Type of Change:
**Delete irrelevant options.**

- Quality Assurance

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
Tests were added to check if email was sent for new mentorship relation requests.

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [x] Update requirements.txt

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/mentorship_relation/test_api_create_request.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import unittest
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_create_request.py]: this test file is already created at `tests/mentorship_relation/test_api_send_request.py`. Can you move the test there, to avoid having tests from the same context in separate files?

File: tests/mentorship_relation/test_api_create_request.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                                            headers=get_test_request_header(self.first_user.id),
+                                            json={
+                                                "mentor_id": self.first_user.id,
+                                                "mentee_id": self.second_user.id,
+                                                "end_date": self.end_date_example.timestamp(),
+                                                "notes": self.notes_example
+                                            })
+
+                self.assertEqual(200, response.status_code)
+                self.assertEqual(1, len(outbox))
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_create_request.py]: you can add these lines 34 and 35 to the tests already in `tests/mentorship_relation/test_api_send_request.py`

File: app/api/email_utils.py [hunk 3]
```diff
@@ -149,3 +149,5 @@ def send_email_new_request(user_sender, user_recipient, notes, sender_role):
     )
     subject = "Mentorship System - You have got new relation request"
     send_email(user_recipient.email, subject, html)
+
```
Comment (isabelcosta) [inline app/api/email_utils.py]: can you undo these changes? just to keep the changes to test module?

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@gaurivn I requested some changes, let me know if you have any doubt or disagree with any of the changes :)

**Comment by gaurivn** [comment]
Sure @isabelcosta thank you, I will make the changes asap.

**Comment by gaurivn** [comment]
@isabelcosta , I made the changes, kindly review them and please tell if I have to make further changes.

File: tests/mentorship_relation/test_api_send_request.py [hunk 4]
```diff
@@ -45,6 +53,8 @@ def test_fail_send_request_bad_mentor_id(self):
                                            headers=auth_header, content_type='application/json',
                                            data=json.dumps(test_payload))
         self.assertEqual(404, actual_response.status_code)
+	self.assertEqual(1, len(outbox))
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_send_request.py]: is this well indented?

File: tests/mentorship_relation/test_api_send_request.py [hunk 5]
```diff
@@ -62,9 +72,45 @@ def test_fail_send_request_bad_user_id(self):
                                            headers=auth_header, content_type='application/json',
                                            data=json.dumps(test_payload))
         self.assertEqual(400, actual_response.status_code)
+	self.assertEqual(1, len(outbox))
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_send_request.py]: is this well indented?

File: app/api/email_utils.py (hunk 3 above)
Comment (isabelcosta) [reply app/api/email_utils.py]: can you remove these changes? I can merge them but if you could take these off as well. This seems like style format changes, that can be done in another PR

**Review by isabelcosta — CHANGES_REQUESTED** [review]
there is left just some indentation fixes @gaurivn

**Comment by gaurivn** [comment]
@isabelcosta , when I checked in gedit editor, it shows proper indentation. I thought it was because of mixing tabs and spaces. I had removed the extra lines under email_utils.py, got a bit confused with the commit conflicts. Can you please check once again and tell if there are any changes I can make.

**Comment by gaurivn** [comment]
@isabelcosta , it is fixed now, I used spaces instead of tab, please review.

File: tests/mentorship_relation/test_api_send_request.py (hunk 5 above)
Comment (gaurivn) [reply tests/mentorship_relation/test_api_send_request.py]: @isabelcosta , I have made the required indentation changes after removing unnecessary commits, please review and suggest changes if any.

File: tests/mentorship_relation/test_api_send_request.py [hunk 6]
```diff
@@ -14,6 +15,12 @@
 class TestSendRequestApi(MentorshipRelationBaseTestCase):
     def setUp(self):
         super(TestSendRequestApi, self).setUp()
+	self.notes_example = 'description of a good mentorship relation'
```
Comment (isabelcosta) [inline tests/mentorship_relation/test_api_send_request.py]: I think this is not well indented and that is why TravisCI is failing

**Review by isabelcosta — CHANGES_REQUESTED** [review]
Can you please check again code indentation @gaurivn :)

File: tests/mentorship_relation/test_api_send_request.py (hunk 6 above)
Comment (gaurivn) [reply tests/mentorship_relation/test_api_send_request.py]: I indented it properly, replaced tab with proper spaces.

File: tests/mentorship_relation/test_api_send_request.py [hunk 7]
```diff
@@ -29,9 +36,10 @@ def test_fail_send_request_bad_mentee_id(self):
                                            headers=auth_header, content_type='application/json',
                                            data=json.dumps(test_payload))
         self.assertEqual(404, actual_response.status_code)
+        self.assertEqual(1, len(outbox))
```
Comment (sumit-badsara) [inline tests/mentorship_relation/test_api_send_request.py:39]: `outbox` isn't defined here.

**Review by sumit-badsara — CHANGES_REQUESTED** [review]
@gaurivn  `outbox` isn't imported/defined anywhere, which is causing the travis ci failure. Fix that first, and for indentation (maybe use some linter)?

File: tests/mentorship_relation/test_api_send_request.py (hunk 7 above)
Comment (isabelcosta) [reply tests/mentorship_relation/test_api_send_request.py:39]: @gaurivn can you please define how to get the outbox from the user email

Comment (gaurivn) [reply tests/mentorship_relation/test_api_send_request.py:39]: @isabelcosta @BadduCoder, can I remove the outbox line

Comment (isabelcosta) [reply tests/mentorship_relation/test_api_send_request.py:39]: @gaurivn I did a quick search on google to try to figure out how to use the `outbox` concept, and I found this in the official documentation of Flask-Mail: https://pythonhosted.org/Flask-Mail/#unit-tests-and-suppressing-emails
Can you see if you can fix the `outbox` usage in this test, before deciding to not use this?

This issue goal is about using something in the code to figure out if the email is being sent, and I think this can be done through this `outbox` concept

**Comment by devkapilbansal** [comment]
@gaurivn can you please make the necessary changes?

**Comment by vj-codes** [comment]
@gaurivn closing due to inactivity
Thank you for your contribution :)
